src/util/dirty.cpp

```diff
@@ -13,47 +17,166 @@
 
 #include <faabric/util/crash.h>
 #include <faabric/util/dirty.h>
+#include <faabric/util/locks.h>
 #include <faabric/util/memory.h>
 #include <faabric/util/testing.h>
 #include <faabric/util/timing.h>
 
 namespace faabric::util {
 
-// This singleton is needed to contain the different singleton
-// instances. We can't make them all static variables in the function.
-class DirtyTrackerSingleton
+/**
+ * Wrapper around the actual bookkeeping behind dirty tracking. This is
+ * deliberately not thread-safe as it will only ever be used in TLS for signal
+ * handlers.
+ */
+class DirtyTrackingRecord
 {
   public:
-    SoftPTEDirtyTracker softpte;
-    SegfaultDirtyTracker sigseg;
-    NoneDirtyTracker none;
+    DirtyTrackingRecord() = default;
+
+    virtual void trackRegion(std::span<uint8_t> region)
+    {
+        nPages = faabric::util::getRequiredHostPages(region.size());
+        dirtyFlags = std::vector<char>(nPages, 0);
+        regionBase = region.data();
+        regionTop = region.data() + region.size();
+    }
+
+    virtual void markPage(void* addr)
+    {
+        long pageNum = ((uint8_t*)addr - regionBase) / HOST_PAGE_SIZE;
+        dirtyFlags[pageNum] = 1;
+    }
+
+    virtual bool isInitialised() { return regionTop != nullptr; }
+
+    virtual int getNPages() { return nPages; }
+
+    virtual std::vector<char> getDirtyFlags() { return dirtyFlags; }
+
+    virtual void reset()
+    {
+        nPages = 0;
+        dirtyFlags.clear();
+        regionBase = nullptr;
+        regionTop = nullptr;
+    }
+
+  protected:
+    int nPages = 0;
+
+    // std::vector<bool> here seems to worsen performance by >4x
+    // std::vector<char> seems to be optimal
+    std::vector<char> dirtyFlags;
+
+    uint8_t* regionBase = nullptr;
+    uint8_t* regionTop = nullptr;
 };
 
-DirtyTracker& getDirtyTracker()
+/**
+ * Thread-safe wrapper around dirty tracking data, necessary for use with the
+ * background event thread in the userfaultfd tracker. Although the logic around
+ * the interaction between the event thread and the main thread should in theory
+ * protect against concurrent accesses, we need this to keep TSan happy.
+ */
+class ThreadsafeDirtyTrackingRecord final : public DirtyTrackingRecord
```

Comment:
Very minor, `ThreadSafeDirtyTrackingRecord` feels like a more natural capitalisation to me.